ESQL: silent java.lang.IllegalStateException error

### Description

```
    PUT /gen_rm_0
    {"mappings":{"properties": {"value": {"type": "text", "index_phrases": true},"tag": {"type": "keyword", "store": true, "index": false},"code": {"type": "keyword", "store": true, "ignore_above": 1831, "index_options": "docs", "split_queries_on_whitespace": true},"message": {"type": "histogram"},"mode": {"type": "double", "ignore_malformed": true, "index": true},"price": {"type": "byte"},"rate": {"type": "double", "coerce": false},"status": {"type": "long"},"region": {"type": "date_range"},"ver": {"type": "date_nanos"},"title": {"type": "date", "doc_values": true},"size": {"type": "flattened"},"category": {"type": "float"},"weight": {"type": "short", "doc_values": false, "store": true, "coerce": false},"result": {"type": "date_nanos", "store": true, "ignore_malformed": true}}}}

    POST /gen_rm_0/_bulk?refresh=true
    {"index":{}}
    {"tag":"kw_yJpk","code":"kw_xRKNC","message":{"values":[-58.28,-46.6,-36.36,-20.18,1.31],"counts":[96,98,61,2,91]},"mode":83.56,"rate":-94.69,"status":-529634,"region":{"gte":"2021-09-16T04:47:15.000Z","lte":"2023-04-12T20:17:09.000Z"},"ver":"2023-04-12T19:36:07.234575943Z","title":"2025-05-13T11:53:12.000Z","size":{"key_inW":"val_xrLZg","num":"37","nested_er":{"inner":"val_ojn"}},"category":[55.58,-20.85],"weight":-14509}
    {"index":{}}
    {"tag":"kw_MmydPR","code":"kw_kIekZAIob","message":{"values":[-77.73,-62.81,-25.11],"counts":[23,46,59]},"mode":0.06,"price":24,"rate":-33.2,"status":589716,"region":{"gte":"2024-04-20T20:00:44.000Z","lte":"2025-01-06T17:29:10.000Z"},"ver":"2020-09-05T03:40:20.025388156Z","title":"2024-02-10T01:25:33.000Z","size":{"key_YXe":"val_myPVO","num":"32","nested_VE":{"inner":"val_qUN"}},"category":30.84,"weight":-3993,"result":"2022-01-13T12:03:01.027500495Z"}
    {"index":{}}
    {"value":"delta OgPWQBlxWuxC","tag":"kw_GiSi","code":"kw_mJgvOOhvM","message":{"values":[-76.85,-75.1],"counts":[94,87]},"mode":86.39,"price":-52,"rate":-29.46,"status":-504336,"region":{"gte":"2020-02-08T18:50:39.000Z","lte":"2021-07-10T22:41:29.000Z"},"title":["2021-12-24T03:34:01.000Z","2020-01-20T17:09:03.000Z"],"weight":30338,"result":"2021-09-13T19:41:51.202516203Z"}
    {"index":{}}
    {"value":["bravo RYoJgSKtUJIB","charlie uPjr"],"code":["kw_wQfYqmleHC","kw_dfU"],"message":{"values":[-18.69,0.1,28.68,49.49,57.96,80.8],"counts":[40,59,35,91,76,45]},"mode":-31.77,"price":-45,"rate":34.36,"status":[-857464,-633300],"region":{"gte":"2020-10-17T06:55:46.000Z","lte":"2020-12-11T17:50:45.000Z"},"ver":"2024-01-22T01:39:14.376564501Z","title":"2022-05-06T21:15:52.000Z","size":{"key_hnE":"val_tAVaT","num":"71","nested_iJ":{"inner":"val_zzO"}},"category":13.2,"weight":10406,"result":"2025-03-10T17:01:09.315891996Z"}
    {"index":{}}
    {"value":"charlie QSpdHmUk","tag":"kw_yiPGSMmTII","code":"kw_jZkVsVPdV","message":{"values":[-66.81,-29.83,19.41,66.72,96.11],"counts":[37,72,76,70,36]},"mode":-89.16,"rate":15.4,"status":[646156,513276],"region":{"gte":"2021-10-12T00:51:20.000Z","lte":"2022-09-20T19:22:02.000Z"},"ver":"2022-11-15T16:20:04.777071381Z","title":"2023-08-24T22:08:50.000Z","size":{"key_cxi":"val_JFoBi","num":"42","nested_bC":{"inner":"val_Ynr"}},"category":91.04}
    {"index":{}}
    {"code":"kw_zVF","message":{"values":[-94.18,-53.2,-33.52],"counts":[97,31,17]},"mode":-67.91,"price":-9,"rate":-27.08,"status":233293,"region":{"gte":"2020-09-15T07:07:06.000Z","lte":"2022-07-16T09:10:31.000Z"},"ver":"2025-07-14T06:48:39.942303415Z","title":"2023-12-21T12:07:30.000Z","size":{"key_dFN":"val_knbzH","num":"74"},"weight":17375}
    {"index":{}}
    {"value":"charlie PaSa","tag":"kw_oHGbXX","code":"kw_rkAD","message":{"values":[-39.94,-15.76],"counts":[30,9]},"price":118,"rate":-96.81,"region":[{"gte":"2021-08-23T04:01:54.000Z","lte":"2024-01-18T08:21:01.000Z"},{"gte":"2020-05-03T23:20:48.000Z","lte":"2020-11-25T05:48:42.000Z"}],"ver":"2020-12-10T16:46:24.833841519Z","size":{"key_CVs":"val_nKVZw","num":"19"},"category":0.8,"weight":[29997,-30239],"result":"2025-11-12T23:36:34.274773425Z"}
    {"index":{}}
    {"value":"echo YkKxaRHDwIk","tag":"kw_BWrGkWTK","code":"kw_ZlCRb","message":{"values":[-62.33,-18.26,-3.65,2.44,23.76,32.47],"counts":[55,61,66,6,81,41]},"mode":-37.29,"price":9,"rate":-18.38,"status":151819,"region":{"gte":"2023-01-05T07:54:35.000Z","lte":"2025-12-06T15:03:41.000Z"},"ver":"2020-07-01T23:27:34.631832345Z","title":"2022-08-12T15:29:28.000Z","size":{"key_DEA":"val_ToASM","num":"82","nested_Mp":{"inner":"val_GXQ"}},"category":84.11,"weight":-12890}
    {"index":{}}
    {"value":"bravo OemCzh","tag":"kw_aedAXphLZd","code":"kw_Wzorhs","message":{"values":[-80.45,-75.54,-43.8,2.93,26.36],"counts":[84,80,1,97,77]},"mode":-37.67,"status":-439268,"region":{"gte":"2023-12-05T10:16:14.000Z","lte":"2025-06-12T09:39:14.000Z"},"ver":"2021-08-21T17:57:20.251533330Z","title":["2023-11-05T01:33:18.000Z","2021-12-27T18:21:46.000Z"],"size":{"key_bXO":"val_gsuYH","num":"87"},"category":10.68}
    {"index":{}}
    {"tag":"kw_IAauUHwLCz","mode":-41.56,"rate":32.68,"status":805059,"region":{"gte":"2025-02-28T15:30:25.000Z","lte":"2025-06-25T12:16:27.000Z"},"ver":"2023-12-28T19:16:21.941396123Z","title":"2024-09-08T18:44:45.000Z","size":{"key_JLJ":"val_FfyoI","num":"11"},"category":-55.02,"weight":27805}
    {"index":{}}
    {"tag":"kw_gqkT","code":"kw_SIGZiV","message":{"values":[-8.81,2.93],"counts":[12,6]},"price":74,"status":568695,"region":{"gte":"2022-05-27T07:40:35.000Z","lte":"2023-03-10T06:41:42.000Z"},"ver":"2020-06-28T18:54:23.794735066Z","title":"2022-08-28T11:17:34.000Z","size":{"key_SEl":"val_WiDEo","num":"13","nested_QN":{"inner":"val_XMT"}},"category":-27.76,"weight":12046,"result":"2025-02-07T04:03:06.770489552Z"}
    {"index":{}}
    {"value":"echo kXvxZ","tag":["kw_YNlRBmN","kw_nds"],"code":"kw_vxvQIoGU","message":{"values":[-80.47,-55.85,-10.0,1.07,15.79],"counts":[74,71,23,19,8]},"mode":-66.11,"price":41,"rate":10.91,"ver":["2025-08-03T23:48:05.188768707Z","2023-01-14T14:24:49.667645871Z"],"size":{"key_MwH":"val_gtxRL","num":"49","nested_Dm":{"inner":"val_AGq"}},"category":68.53,"weight":26718,"result":"2022-03-13T23:52:11.033084143Z"}
    {"index":{}}
    {"value":"bravo pVahVofkRJKy","tag":["kw_iNp","kw_kGV"],"code":["kw_VbmVI","kw_TgJRv"],"message":{"values":[-86.53,-45.26,-25.75,-3.03,23.12,47.32],"counts":[46,75,41,63,57,13]},"mode":-31.46,"price":-57,"status":[-349388,-460803],"title":"2025-10-09T05:44:54.000Z","size":{"key_foh":"val_gjhNZ","num":"51","nested_SF":{"inner":"val_Bzv"}},"category":-92.18,"weight":21155}
```

```
    PUT /gen_rm_2
    {"mappings":{"properties": {"mode": {"type": "double", "ignore_malformed": true, "index": true},"duration": {"type": "double", "index": false},"title": {"type": "date", "doc_values": true},"value": {"type": "text", "index_phrases": true, "store": true},"weight": {"type": "short", "doc_values": false, "store": true, "coerce": false},"message": {"type": "token_count", "analyzer": "stop", "index": false},"height": {"type": "text"},"score": {"type": "date_nanos", "doc_values": false},"status": {"type": "long", "doc_values": true, "coerce": false},"name": {"type": "boolean"},"amount": {"type": "boolean", "ignore_malformed": true},"count": {"type": "dense_vector", "dims": 128, "index": false, "element_type": "byte"},"price": {"type": "byte", "doc_values": true, "coerce": false},"level": {"type": "date", "doc_values": false, "index": true},"distance": {"type": "token_count", "analyzer": "simple", "store": true, "null_value": 0}}}}

    POST /gen_rm_2/_bulk?refresh=true
    {"index":{}}
    {"mode":-30.49,"title":"2020-04-11T03:09:04.000Z","value":"charlie QCKtLTZZ","weight":-944,"message":"hello","score":["2020-12-04T18:51:44.313027313Z","2020-05-10T09:56:19.985234466Z"],"status":[964635,737754],"name":false,"amount":false,"count":[-62,-101,21,91,54,35,-97,80,-21,77,-93,-87,53,31,-117,74,-113,63,43,118,9,-7,-63,-113,50,-9,37,49,-61,37,49,-60,-103,40,-104,-42,112,-38,19,20,60,-30,6,-93,-113,-60,-71,53,70,-108,67,106,94,-4,87,-116,116,72,108,24,-59,-114,-56,-17,-61,-55,97,65,-111,-26,77,115,101,105,51,-43,-64,-38,70,-49,-97,-106,70,52,6,13,-40,-99,-127,68,-64,-57,22,41,114,-56,-44,-90,-71,22,-6,-49,-59,23,-19,89,-15,78,113,48,-13,-16,65,2,-73,92,5,62,-87,-29,123,116,-97,-88,-99,116,92,-90],"price":-21,"level":"2022-02-25T01:47:42.000Z","distance":"single"}
    {"index":{}}
    {"mode":-64.78,"duration":14.77,"title":"2021-08-19T02:42:36.000Z","value":"charlie eKTULjV","weight":-19445,"message":"hello","height":"alpha PVHavLMj","status":900396,"name":true,"amount":false,"count":[31,-127,-64,-70,-76,8,-71,-103,68,25,-95,-109,-82,51,40,-77,-99,100,-17,-61,-40,-91,-30,-51,-110,25,64,-50,117,52,-70,-68,46,-120,-99,-99,124,20,-5,33,105,30,-18,37,15,-20,-64,117,105,-77,-10,80,-100,-41,3,67,45,-15,40,97,-31,122,10,-60,-7,23,-109,32,-83,79,-67,-55,-3,115,-71,100,71,-39,121,78,-70,21,92,-87,26,-52,-81,56,-77,51,-124,25,104,100,-43,28,-13,65,-22,-82,9,-30,-125,61,-1,-92,3,-71,112,-49,51,53,-53,-25,70,-16,-63,59,-97,-103,-13,-112,-55,-88,-89,106,106,-73],"price":-46,"level":"2021-03-28T06:14:16.000Z","distance":"single"}
    {"index":{}}
    {"mode":[86.1,-70.6],"duration":-69.68,"value":"charlie NafwMndYpFd","weight":-1672,"message":"single","height":"bravo SvS","score":"2023-12-21T21:48:07.868391962Z","status":347556,"name":false,"count":[10,-29,-64,-46,-126,36,52,-122,13,-101,124,24,8,126,119,74,16,-83,-40,-36,-107,58,96,-4,-40,89,-66,-69,119,-32,-121,-124,-125,-56,127,-109,49,56,68,75,33,95,-66,-82,-25,-2,114,112,57,35,-3,-121,-18,11,10,-70,61,26,100,-81,56,40,117,-54,-116,-111,-9,-105,-16,-34,-91,-53,126,113,-84,98,-31,-59,-64,-14,-20,79,-101,2,70,-36,-111,42,126,55,-115,55,-30,43,32,90,10,89,93,113,95,38,60,-7,80,-45,22,-4,112,-127,41,91,-39,53,-86,76,16,-58,-15,34,39,-71,-67,98,59,-77,-36,0],"price":103,"level":"2025-11-04T12:09:56.000Z","distance":"quick brown fox jumped"}
    {"index":{}}
    {"mode":63.0,"duration":26.34,"value":["echo bJTD","bravo Setmsk"],"weight":-5071,"message":"single","height":["bravo rLJzT","bravo owU"],"status":-318337,"name":false,"amount":true,"count":[-45,-30,-87,-97,70,-118,51,4,-43,-114,7,125,-1,-116,-10,8,-54,-30,-21,-94,-16,-59,90,-43,113,90,56,17,51,76,-44,35,-20,-121,67,-73,105,79,-93,-14,87,-100,-18,5,-83,-6,-76,-111,124,15,23,82,-25,120,28,-62,-115,52,-79,4,65,48,-97,110,37,70,12,60,72,66,-80,102,8,-98,-49,-119,-8,-33,69,4,-106,12,-11,39,9,-104,-44,-20,87,37,-63,-1,122,37,70,94,-49,-62,33,69,-63,-59,80,75,-66,108,82,-112,-60,29,-103,-109,-65,45,-5,16,-21,110,44,12,-102,112,63,51,-124,-57,-25,-64],"level":"2020-05-17T05:11:08.000Z","distance":"one two three"}
    {"index":{}}
    {"mode":20.45,"title":"2022-04-28T19:36:27.000Z","value":"delta jRKeYRxR","weight":-18945,"message":"one two three","height":"echo EDxwBxz","score":"2025-08-17T04:58:21.547057289Z","name":false,"amount":true,"price":-40,"distance":"single"}
```

And query `from gen_rm_2,gen_rm_0` this error is logged in logs, but the result of the query is still returned:

```
[ERROR][o.e.c.o.Driver           ] [runTask-0] Error running driver [data] java.lang.IllegalStateException: region[BlockDocValuesReader.DateRangeDocValuesReader]: org.elasticsearch.compute.data.LongRangeArrayBlock@ec3c9a60 has [14] positions instead of [13]
        at org.elasticsearch.compute.lucene.read.ValuesSourceReaderOperator.sanityCheckBlock(ValuesSourceReaderOperator.java:529)
        at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.loadFromSingleLeaf(ValuesFromSingleReader.java:120)
        at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.load(ValuesFromSingleReader.java:59)
        at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:46)
        at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:26)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:342)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:318)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:194)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:179)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator.getOutput(AbstractPageMappingToIteratorOperator.java:153)
        at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:323)
        at org.elasticsearch.compute.operator.Driver.run(Driver.java:194)
        at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:506)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.compute.operator.DriverScheduler$1.doRun(DriverScheduler.java:57)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:35)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1114)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
        at java.base/java.lang.Thread.run(Thread.java:1474)

[WARN ][o.e.x.e.a.EsqlResponseListener] [runTask-0] partial failure at path: /_query, params: {format=txt, error_trace=true} shard [[_na_][gen_rm_0][0]], reason [java.lang.IllegalStateException: region[BlockDocValuesReader.DateRangeDocValuesReader]: org.elasticsearch.compute.data.LongRangeArrayBlock@ec3c9a60 has [14] positions instead of [13]
        at org.elasticsearch.compute.lucene.read.ValuesSourceReaderOperator.sanityCheckBlock(ValuesSourceReaderOperator.java:529)
        at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.loadFromSingleLeaf(ValuesFromSingleReader.java:120)
        at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.load(ValuesFromSingleReader.java:59)
        at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:46)
        at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:26)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:342)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:318)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:194)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:179)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator.getOutput(AbstractPageMappingToIteratorOperator.java:153)
        at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:323)
        at org.elasticsearch.compute.operator.Driver.run(Driver.java:194)
        at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:506)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.compute.operator.DriverScheduler$1.doRun(DriverScheduler.java:57)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:35)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1114)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
        at java.base/java.lang.Thread.run(Thread.java:1474)
], cause [java.lang.IllegalStateException: region[BlockDocValuesReader.DateRangeDocValuesReader]: org.elasticsearch.compute.data.LongRangeArrayBlock@ec3c9a60 has [14] positions instead of [13]
        at org.elasticsearch.compute.lucene.read.ValuesSourceReaderOperator.sanityCheckBlock(ValuesSourceReaderOperator.java:529)
        at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.loadFromSingleLeaf(ValuesFromSingleReader.java:120)
        at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.load(ValuesFromSingleReader.java:59)
        at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:46)
        at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:26)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:342)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:318)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:194)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:179)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator.getOutput(AbstractPageMappingToIteratorOperator.java:153)
        at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:323)
        at org.elasticsearch.compute.operator.Driver.run(Driver.java:194)
        at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:506)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.compute.operator.DriverScheduler$1.doRun(DriverScheduler.java:57)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:35)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1114)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
        at java.base/java.lang.Thread.run(Thread.java:1474)
]
        at org.elasticsearch.xpack.esql.plugin.DataNodeRequestSender.selectFailures(DataNodeRequestSender.java:275)
        at org.elasticsearch.xpack.esql.plugin.DataNodeRequestSender.lambda$startComputeOnDataNodes$0(DataNodeRequestSender.java:160)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$MappedActionListener.onResponse(ActionListenerImplementations.java:105)
        at org.elasticsearch.xpack.esql.plugin.ComputeListener.lambda$new$0(ComputeListener.java:37)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingFailureActionListener.onResponse(ActionListenerImplementations.java:233)
        at org.elasticsearch.compute.EsqlRefCountingListener.lambda$new$0(EsqlRefCountingListener.java:31)
        at org.elasticsearch.base@9.5.0-SNAPSHOT/org.elasticsearch.core.AbstractRefCounted$1.closeInternal(AbstractRefCounted.java:125)
        at org.elasticsearch.base@9.5.0-SNAPSHOT/org.elasticsearch.core.AbstractRefCounted.decRef(AbstractRefCounted.java:77)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.support.RefCountingRunnable.close(RefCountingRunnable.java:122)
        at org.elasticsearch.base@9.5.0-SNAPSHOT/org.elasticsearch.core.Releasables$4.close(Releasables.java:181)
        at org.elasticsearch.base@9.5.0-SNAPSHOT/org.elasticsearch.core.Releasables.close(Releasables.java:39)
        at org.elasticsearch.base@9.5.0-SNAPSHOT/org.elasticsearch.core.Releasables.closeExpectNoException(Releasables.java:72)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$2.run(ActionListenerImplementations.java:51)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:234)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListener$3.onResponse(ActionListener.java:432)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onResponse(ActionListenerImplementations.java:350)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:198)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onResponse(ActionListenerImplementations.java:350)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onResponse(ActionListenerImplementations.java:198)
        at org.elasticsearch.xpack.esql.plugin.ComputeListener.lambda$acquireCompute$2(ComputeListener.java:63)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onResponse(ActionListener.java:274)
        at org.elasticsearch.xpack.esql.plugin.DataNodeRequestSender$1.onFailure(DataNodeRequestSender.java:336)
        at org.elasticsearch.xpack.esql.plugin.DataNodeComputeHandler$1.lambda$sendRequest$0(DataNodeComputeHandler.java:156)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:282)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$RunAfterActionListener.onFailure(ActionListenerImplementations.java:321)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListener$3.onFailure(ActionListener.java:442)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
        at org.elasticsearch.compute.EsqlRefCountingListener.lambda$new$0(EsqlRefCountingListener.java:29)
        at org.elasticsearch.base@9.5.0-SNAPSHOT/org.elasticsearch.core.AbstractRefCounted$1.closeInternal(AbstractRefCounted.java:125)
        at org.elasticsearch.base@9.5.0-SNAPSHOT/org.elasticsearch.core.AbstractRefCounted.decRef(AbstractRefCounted.java:77)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.support.RefCountingRunnable.close(RefCountingRunnable.java:122)
        at org.elasticsearch.base@9.5.0-SNAPSHOT/org.elasticsearch.core.Releasables$4.close(Releasables.java:181)
        at org.elasticsearch.base@9.5.0-SNAPSHOT/org.elasticsearch.core.Releasables.close(Releasables.java:39)
        at org.elasticsearch.base@9.5.0-SNAPSHOT/org.elasticsearch.core.Releasables.closeExpectNoException(Releasables.java:72)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$2.run(ActionListenerImplementations.java:51)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListener$1.lambda$onFailure$0(ActionListener.java:243)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListener$1.onFailure(ActionListener.java:243)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListener$3.onFailure(ActionListener.java:442)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onFailure(ActionListenerImplementations.java:360)
        at org.elasticsearch.compute.EsqlRefCountingListener.lambda$acquire$1(EsqlRefCountingListener.java:40)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.acceptException(ActionListenerImplementations.java:202)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onFailure(ActionListenerImplementations.java:207)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$RunBeforeActionListener.onFailure(ActionListenerImplementations.java:360)
        at org.elasticsearch.xpack.esql.plugin.ComputeListener.lambda$acquireAvoid$1(ComputeListener.java:50)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.acceptException(ActionListenerImplementations.java:202)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$DelegatingResponseActionListener.onFailure(ActionListenerImplementations.java:207)
        at org.elasticsearch.xpack.esql.plugin.ComputeListener.lambda$acquireCompute$3(ComputeListener.java:66)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:282)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeAcceptException(ActionListenerImplementations.java:78)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations.safeOnFailure(ActionListenerImplementations.java:89)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.DelegatingActionListener.onFailure(DelegatingActionListener.java:32)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:54)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleException(TransportService.java:1755)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1535)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1669)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.transport.TransportService$DirectResponseChannel$2.doRun(TransportService.java:1649)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:35)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1114)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
        at java.base/java.lang.Thread.run(Thread.java:1474)
Caused by: java.lang.IllegalStateException: region[BlockDocValuesReader.DateRangeDocValuesReader]: org.elasticsearch.compute.data.LongRangeArrayBlock@ec3c9a60 has [14] positions instead of [13]
        at org.elasticsearch.compute.lucene.read.ValuesSourceReaderOperator.sanityCheckBlock(ValuesSourceReaderOperator.java:529)
        at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.loadFromSingleLeaf(ValuesFromSingleReader.java:120)
        at org.elasticsearch.compute.lucene.read.ValuesFromSingleReader.load(ValuesFromSingleReader.java:59)
        at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:46)
        at org.elasticsearch.compute.lucene.read.ValuesReader.next(ValuesReader.java:26)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:342)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$AppendBlocksIterator.next(AbstractPageMappingToIteratorOperator.java:318)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:194)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator$RuntimeTrackingIterator.next(AbstractPageMappingToIteratorOperator.java:179)
        at org.elasticsearch.compute.operator.AbstractPageMappingToIteratorOperator.getOutput(AbstractPageMappingToIteratorOperator.java:153)
        at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:323)
        at org.elasticsearch.compute.operator.Driver.run(Driver.java:194)
        at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:506)
        at org.elasticsearch.server@9.5.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.compute.operator.DriverScheduler$1.doRun(DriverScheduler.java:57)
        ... 7 more
```

--- a/docs/changelog/147153.yaml
+++ b/docs/changelog/147153.yaml
@@ -1,0 +1,6 @@
+area: ES|QL
+issues:
+ - 146380
+pr: 147153
+summary: Fix `DateRangeDocValuesReader` to `appendNull` for empty positions
+type: bug

--- a/server/src/main/java/org/elasticsearch/index/mapper/blockloader/docvalues/DateRangeDocValuesLoader.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/blockloader/docvalues/DateRangeDocValuesLoader.java
@@ -75,15 +75,19 @@ public class DateRangeDocValuesLoader extends BlockDocValuesReader.DocValuesBloc
                     } else {
                         BytesRef ref = docValues.docValues().binaryValue();
                         var ranges = BinaryRangeUtil.decodeLongRanges(ref);
-                        for (var range : ranges) {
-                            lastDoc = doc;
-                            this.docId = doc;
-                            // While the index stores ranges with fully-inclusive bounds [from, to], in ESQL we always
-                            // represent and use date ranges as half-open [from, to), so we add 1 to the upper bound here.
-                            long from = (long) range.getFrom();
-                            long toIndex = (long) range.getTo();
-                            builder.from().appendLong(from);
-                            builder.to().appendLong(toIndex + 1);
+                        if (ranges.isEmpty()) {
+                            builder.appendNull();
+                        } else {
+                            for (var range : ranges) {
+                                lastDoc = doc;
+                                this.docId = doc;
+                                // While the index stores ranges with fully-inclusive bounds [from, to], in ESQL we always
+                                // represent and use date ranges as half-open [from, to), so we add 1 to the upper bound here.
+                                long from = (long) range.getFrom();
+                                long toIndex = (long) range.getTo();
+                                builder.from().appendLong(from);
+                                builder.to().appendLong(toIndex + 1);
+                            }
                         }
                     }
                 }

--- a/server/src/main/java/org/elasticsearch/index/mapper/blockloader/docvalues/DateRangeDocValuesLoader.java
+++ b/server/src/main/java/org/elasticsearch/index/mapper/blockloader/docvalues/DateRangeDocValuesLoader.java
@@ -70,6 +70,8 @@ public class DateRangeDocValuesLoader extends BlockDocValuesReader.DocValuesBloc
                     if (doc < lastDoc) {
                         throw new IllegalStateException("docs within same block must be in order");
                     }
+                    lastDoc = doc;
+                    this.docId = doc;
                     if (false == docValues.docValues().advanceExact(doc)) {
                         builder.appendNull();
                     } else {
@@ -77,17 +79,21 @@ public class DateRangeDocValuesLoader extends BlockDocValuesReader.DocValuesBloc
                         var ranges = BinaryRangeUtil.decodeLongRanges(ref);
                         if (ranges.isEmpty()) {
                             builder.appendNull();
+                        } else if (ranges.size() == 1) {
+                            var range = ranges.get(0);
+                            // While the index stores ranges with fully-inclusive bounds [from, to], in ESQL we always
+                            // represent and use date ranges as half-open [from, to), so we add 1 to the upper bound here.
+                            builder.from().appendLong((long) range.getFrom());
+                            builder.to().appendLong((long) range.getTo() + 1);
                         } else {
+                            builder.from().beginPositionEntry();
+                            builder.to().beginPositionEntry();
                             for (var range : ranges) {
-                                lastDoc = doc;
-                                this.docId = doc;
-                                // While the index stores ranges with fully-inclusive bounds [from, to], in ESQL we always
-                                // represent and use date ranges as half-open [from, to), so we add 1 to the upper bound here.
-                                long from = (long) range.getFrom();
-                                long toIndex = (long) range.getTo();
-                                builder.from().appendLong(from);
-                                builder.to().appendLong(toIndex + 1);
+                                builder.from().appendLong((long) range.getFrom());
+                                builder.to().appendLong((long) range.getTo() + 1);
                             }
+                            builder.from().endPositionEntry();
+                            builder.to().endPositionEntry();
                         }
                     }
                 }

--- a/server/src/test/java/org/elasticsearch/index/mapper/DateRangeDocValuesLoaderTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/DateRangeDocValuesLoaderTests.java
@@ -1,0 +1,70 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.document.BinaryDocValuesField;
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.tests.index.RandomIndexWriter;
+import org.elasticsearch.common.unit.ByteSizeValue;
+import org.elasticsearch.index.mapper.blockloader.docvalues.DateRangeDocValuesLoader;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+public class DateRangeDocValuesLoaderTests extends ESTestCase {
+
+    private static final String FIELD_NAME = "test_date_range";
+
+    /**
+     * Verifies that a document whose binary doc value encodes zero ranges produces a null entry in the
+     * loaded block rather than silently skipping the position, which would cause the block to have
+     * fewer positions than expected and trigger a sanity-check failure in ValuesSourceReaderOperator.
+     */
+    public void testEmptyRangesAppendsNull() throws IOException {
+        // Encode a single valid inclusive range [1000, 2000].
+        var oneRange = BinaryRangeUtil.encodeLongRanges(Set.of(new RangeFieldMapper.Range(RangeType.DATE, 1000L, 2000L, true, true)));
+        // Encode zero ranges — this is the edge case that triggered the bug.
+        var emptyRanges = BinaryRangeUtil.encodeLongRanges(Set.of());
+
+        try (Directory dir = newDirectory(); RandomIndexWriter iw = new RandomIndexWriter(random(), dir)) {
+            // Doc 0: one valid range.
+            iw.addDocument(List.of(new BinaryDocValuesField(FIELD_NAME, oneRange)));
+            // Doc 1: binary doc value present but encodes zero ranges — the bug case.
+            iw.addDocument(List.of(new BinaryDocValuesField(FIELD_NAME, emptyRanges)));
+            // Doc 2: no binary doc value at all — the normal null case.
+            iw.addDocument(List.of());
+            iw.forceMerge(1);
+
+            try (DirectoryReader dr = iw.getReader()) {
+                LeafReaderContext ctx = getOnlyLeafReader(dr).getContext();
+                var loader = new DateRangeDocValuesLoader(FIELD_NAME);
+                try (var reader = loader.columnAtATimeReader(ctx).apply(newLimitedBreaker(ByteSizeValue.ofMb(1)))) {
+                    var block = (TestBlock) reader.read(TestBlock.factory(), TestBlock.docs(0, 1, 2), 0, false);
+
+                    assertThat(block.size(), equalTo(3));
+                    // Doc 0: non-null range present — to is stored as upper+1 (half-open [from, to)).
+                    assertThat(block.get(0), notNullValue());
+                    // Doc 1: zero ranges encoded — must produce null, not skip the position.
+                    assertThat(block.get(1), nullValue());
+                    // Doc 2: no doc value — null.
+                    assertThat(block.get(2), nullValue());
+                }
+            }
+        }
+    }
+}

--- a/server/src/test/java/org/elasticsearch/index/mapper/DateRangeDocValuesLoaderTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/DateRangeDocValuesLoaderTests.java
@@ -30,6 +30,38 @@ public class DateRangeDocValuesLoaderTests extends ESTestCase {
     private static final String FIELD_NAME = "test_date_range";
 
     /**
+     * Verifies that a document with multiple stored ranges produces a multi-value entry in the loaded
+     * block, with each range represented as a half-open [from, to) interval.
+     */
+    public void testMultipleRangesProducesMultiValueEntry() throws IOException {
+        // encodeLongRanges sorts by from then to, so [1000,2000] comes before [3000,4000].
+        var twoRanges = BinaryRangeUtil.encodeLongRanges(
+            Set.of(
+                new RangeFieldMapper.Range(RangeType.DATE, 1000L, 2000L, true, true),
+                new RangeFieldMapper.Range(RangeType.DATE, 3000L, 4000L, true, true)
+            )
+        );
+
+        try (Directory dir = newDirectory(); RandomIndexWriter iw = new RandomIndexWriter(random(), dir)) {
+            iw.addDocument(List.of(new BinaryDocValuesField(FIELD_NAME, twoRanges)));
+            iw.forceMerge(1);
+
+            try (DirectoryReader dr = iw.getReader()) {
+                LeafReaderContext ctx = getOnlyLeafReader(dr).getContext();
+                var loader = new DateRangeDocValuesLoader(FIELD_NAME);
+                try (var reader = loader.columnAtATimeReader(ctx).apply(newLimitedBreaker(ByteSizeValue.ofMb(1)))) {
+                    var block = (TestBlock) reader.read(TestBlock.factory(), TestBlock.docs(0), 0, false);
+
+                    assertThat(block.size(), equalTo(1));
+                    // Both ranges stored as inclusive [from, to] are converted to half-open [from, to+1).
+                    // The from-values and to-values are each collected into a list for the multi-value position.
+                    assertThat(block.get(0), equalTo(List.of(List.of(1000L, 3000L), List.of(2001L, 4001L))));
+                }
+            }
+        }
+    }
+
+    /**
      * Verifies that a document whose binary doc value encodes zero ranges produces a null entry in the
      * loaded block rather than silently skipping the position, which would cause the block to have
      * fewer positions than expected and trigger a sanity-check failure in ValuesSourceReaderOperator.

--- a/server/src/test/java/org/elasticsearch/index/mapper/DateRangeDocValuesLoaderTests.java
+++ b/server/src/test/java/org/elasticsearch/index/mapper/DateRangeDocValuesLoaderTests.java
@@ -23,7 +23,6 @@ import java.util.List;
 import java.util.Set;
 
 import static org.hamcrest.Matchers.equalTo;
-import static org.hamcrest.Matchers.notNullValue;
 import static org.hamcrest.Matchers.nullValue;
 
 public class DateRangeDocValuesLoaderTests extends ESTestCase {
@@ -57,8 +56,8 @@ public class DateRangeDocValuesLoaderTests extends ESTestCase {
                     var block = (TestBlock) reader.read(TestBlock.factory(), TestBlock.docs(0, 1, 2), 0, false);
 
                     assertThat(block.size(), equalTo(3));
-                    // Doc 0: non-null range present — to is stored as upper+1 (half-open [from, to)).
-                    assertThat(block.get(0), notNullValue());
+                    // Doc 0: [1000, 2000] inclusive stored as half-open [1000, 2001) in the block.
+                    assertThat(block.get(0), equalTo(List.of(1000L, 2001L)));
                     // Doc 1: zero ranges encoded — must produce null, not skip the position.
                     assertThat(block.get(1), nullValue());
                     // Doc 2: no doc value — null.

--- a/test/framework/src/main/java/org/elasticsearch/index/mapper/TestBlock.java
+++ b/test/framework/src/main/java/org/elasticsearch/index/mapper/TestBlock.java
@@ -995,14 +995,21 @@ public class TestBlock implements BlockLoader.Block {
             assert fromBlock.size() == toBlock.size();
             var values = new ArrayList<>(fromBlock.size());
             for (int i = 0; i < fromBlock.size(); i++) {
-                values.add(List.of(fromBlock.values.get(i), toBlock.values.get(i)));
+                Object f = fromBlock.values.get(i);
+                if (f == null) {
+                    values.add(null);
+                } else {
+                    values.add(List.of(f, toBlock.values.get(i)));
+                }
             }
             return new TestBlock(values);
         }
 
         @Override
         public BlockLoader.Builder appendNull() {
-            throw new UnsupportedOperationException();
+            from.appendNull();
+            to.appendNull();
+            return this;
         }
 
         @Override

--- a/x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/generative/GenerativeRandomMappingRestTest.java
+++ b/x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/generative/GenerativeRandomMappingRestTest.java
@@ -24,9 +24,7 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
-import java.util.Set;
 import java.util.regex.Pattern;
-import java.util.stream.Collectors;
 
 /**
  * Generative test that creates random Elasticsearch indices with varied mappings
@@ -59,18 +57,6 @@ public abstract class GenerativeRandomMappingRestTest extends GenerativeRestTest
     }).build();
 
     private static volatile List<GeneratedIndex> generatedIndices;
-
-    private static final Set<String> ADDITIONAL_ALLOWED_ERRORS = Set.of(
-        // DateRangeDocValuesReader produces blocks with incorrect position counts when reading date_range
-        // fields with multi-value documents. The mismatch causes an IllegalStateException at the compute layer,
-        // surfaced as partial results. https://github.com/elastic/elasticsearch/issues/146380
-        "RangeArrayBlock.*has \\[\\d+\\] positions instead of"
-    );
-
-    private static final Set<Pattern> ADDITIONAL_ALLOWED_PATTERNS = ADDITIONAL_ALLOWED_ERRORS.stream()
-        .map(x -> ".*" + x + ".*")
-        .map(x -> Pattern.compile(x, Pattern.DOTALL))
-        .collect(Collectors.toSet());
 
     private static final Pattern CANNOT_LOAD_BLOCKS_WITHOUT_DOC_VALUES = Pattern.compile(
         ".*Cannot load blocks without doc values.*",
@@ -235,11 +221,6 @@ public abstract class GenerativeRandomMappingRestTest extends GenerativeRestTest
 
     private static boolean isAdditionalAllowedError(String errorMessage, String query) {
         if (errorMessage == null) return false;
-        for (Pattern p : ADDITIONAL_ALLOWED_PATTERNS) {
-            if (isAllowedError(errorMessage, p)) {
-                return true;
-            }
-        }
         // RangeFieldMapper throws "Cannot load blocks without doc values" for *_range fields
         // with doc_values:false. https://github.com/elastic/elasticsearch/issues/146527
         if (isAllowedError(errorMessage, CANNOT_LOAD_BLOCKS_WITHOUT_DOC_VALUES) && hasRangeFieldType()) {

